state/presence/presence.go

```diff
@@ -757,6 +861,174 @@ func (p *Pinger) ping() (err error) {
+		}},
+		// Filter out any keys that have only a single sequence
+		// representing them
+		// Note: indexing is from 0, you can set this to 2 if you wanted
+		// to only bother pruning sequences that have >2 entries.
+		// This mostly helps the 'nothing to do' case, dropping the time
+		// to realize there are no sequences to be removed from 36ms,
+		// down to 15ms with 3500 keys.
+		{"$match": bson.M{"seqs.1": bson.M{"$exists": 1}}},
+	})
+	pipe.Batch(1600)
+	return pipe.Iter()
+}
+
+// queueRemoval includes this sequence as one that has been superseded
+func (p *BeingPruner) queueRemoval(seq int64) {
+	p.toRemove = append(p.toRemove, docIDInt64(p.modelUUID, seq))
+}
+
+// flushRemovals makes sure that we've applied all desired removals
+func (p *BeingPruner) flushRemovals() error {
+	if len(p.toRemove) == 0 {
+		return nil
+	}
+	matched, err := p.beingsC.RemoveAll(bson.M{"_id": bson.M{"$in": p.toRemove}})
+	p.toRemove = p.toRemove[:0]
+	if matched.Removed > 0 {
+		p.removedCount += uint64(matched.Removed)
+	}
+	return err
+}
+
+func (p *BeingPruner) removeOldPings() error {
+	// now and now-period are both considered active slots, so we don't
+	// touch those. We also leave 2 more slots around
+	startTime := time.Now()
+	// TODO(jam): 2017-04-18 startCount may not be appropriate, because
+	// Count() includes all models, not just the one we are pruning
+	startCount, err := p.pingsC.Count()
+	if err != nil {
+		return err
+	}
+	logger.Tracef("pruning %q for %q starting with %d docs",
+		p.pingsC.Name, p.modelUUID, startCount)
+	s := timeSlot(time.Now(), p.delta)
+	oldSlot := s - 3*period
+	res, err := p.pingsC.RemoveAll(bson.D{{"_id", bson.RegEx{"^" + p.modelUUID, ""}},
+					     {"slot", bson.M{"$lt": oldSlot}} })
+	if err != nil && err != mgo.ErrNotFound {
+		logger.Errorf("error removing old entries from %q: %v", p.pingsC.Name, err)
+		return err
+	}
+	endCount, _ := p.pingsC.Count()
+	logger.Debugf("pruned %q for %q (with %d docs) of %d old pings down to %d docs in %v",
+		p.pingsC.Name, p.modelUUID, startCount, res.Removed, endCount, time.Since(startTime))
+	return nil
+}
+
+func (p *BeingPruner) removeUnusedBeings() error {
+	var keyInfo collapsedBeingsInfo
+	// TODO(jam): 2017-04-18 startCount may not be appropriate, because
+	// Count() includes all models, not just the one we are pruning
+	startCount, err := p.beingsC.Count()
+	if err != nil {
+		return err
+	}
+	logger.Tracef("pruning %q for %q starting with %d docs",
+		p.beingsC.Name, p.modelUUID, startCount)
+	startTime := time.Now()
+	keyCount := 0
+	seqCount := 0
+	iter := p.iterKeys()
+	for iter.Next(&keyInfo) {
+		keyCount += 1
+		// Find the max
+		maxSeq := int64(-1)
+		for _, seq := range keyInfo.Seqs {
+			if seq > maxSeq {
+				maxSeq = seq
+			}
+		}
+		// Queue everything < max to be deleted
+		for _, seq := range keyInfo.Seqs {
+			seqCount++
+			if seq >= maxSeq {
+				// It shouldn't be possible to be > at this point
+				continue
+			}
+			p.queueRemoval(seq)
```

Comment:
It would be cleaner to have a separate bulk remover type (similar to what we did in juju/txn). This separates that concern from the main pruner type, moving the associated state out of the pruner. The code would be clearer and less bug-prone.